bpf-filter/filter.c

```diff
@@ -0,0 +1,13 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#include <linux/bpf.h>
+#include <bpf/bpf_helpers.h>
+#include <linux/pkt_cls.h>
+
+SEC("classifier")
+int  cls_filter(struct __sk_buff *skb)
+{
+    return TC_H_MAKE(1, 10);
```

Comment:
The return value is the action or verdict (e.g. TC_ACT_OK or TC_ACT_SHOT).

The class id is set via skb->priority field.

Like: https://github.com/xdp-project/xdp-cpumap-tc/blob/master/src/tc_classify_kern.c#L226

Reply:
This is only true if you use the 'direct-action' parameter to cls_bpf, which you're doing in xdp-cpumap-tc, but Frey is not (in the version of the script that is in this PR). If direct-action is *not* set, the return value will be used as the class ID. If the direct-action option *is* set, the classid can instead be set using `skb->tc_classid` from BPF. This field is initialised to the default classid set on the command line (using 'classid' or 'flowid' when installing the filter), and if the BPF program changes it, the result will be OR'ed together with the major from the command line (meaning you can only set the minor number). This is all in `cls_bpf_classify()` in cls_bpf.c

Also, the fact that you can also set the class for HTB using `skb->priority` is specific to that qdisc; for instance, for sch_cake we use skb->priority to select the tin, while the classifier value is used to select the flow...

Reply:
Uh, and adding to the madness, the verifier will rewrite stores to `skb->tc_classid` to only be two bytes, so you **can** only set the minor number from BPF (which I guess explains the OR'ing after the fact to get the full ID). See `bpf_convert_ctx_access()` in net/core/filter.c (search for `tc_classid`).

Reply:
OMG - this sounds very confusing.
Thanks for educating us, including me, as I didn't know these details!

Reply:
Yeah, I agree, it's incredibly confusing! You're welcome :)

Reply:
That's incredibly confusing. I should have spent more time reading through the TC_H_MAKE macro. It went entirely past me that it didn't shift the major number before it merged the numbers. When I used skb->priority and skb->tc_classid, I honestly thought the numbers were decimal, not hex. It also went past me that the tc_classid only deals with 16-bit numbers. It sort of feels like these classes could have a more common interface to make it easier to understand.

Reply:
Yup, I agree it's incredibly confusing. And really, the only way to figure this out is to try to decipher the code, or ask someone who already knows. The latter is obviously much more efficient, which is why I keep saying that you can just fire off an email to Jesper and myself when you get stuck on something - we're here to help :)